### PR TITLE
fix(realtime): centralize room grammar and revocation kicks

### LINKED ISSUE #2158: realtime: socket room membership vs DB permissions — three sync mechanisms for one fact, and broadcast-audience's hand-maintained room grammar is already stale
## What

Who-can-receive-events exists twice: Postgres permissions (`packages/lib/src/permissions/`) and in-memory Socket.IO room membership (`apps/realtime/src/socket-registry.ts:11-22`, `presence-tracker.ts:23-31` — per-process singletons; a second realtime instance would be a third copy). Keeping them aligned currently takes **three separate mechanisms**:

1. **Push (kick calls)**: web must remember to call `kickUserFromRooms` → `POST /api/kick` (`lib/websocket/socket-utils.ts:1374-1511`, handled by `apps/realtime/src/kick-handler.ts:121-187`) on every revocation path. Call sites are scattered per-route (`api/drives/[driveId]/members/[userId]/route.ts:22-23`, `api/activities/[activityId]/rollback/route.ts:184-196`, …) — any new revocation path that forgets the call leaves stale room membership.
2. **Pull (per-event recheck)**: `apps/realtime/src/per-event-auth.ts:1-14` re-queries Postgres per sensitive event, explicitly to patch "the 'revoked but still in room' attack window" — i.e., the cache is acknowledged untrustworthy, so the DB is consulted anyway (related: closed #422, #424).
3. **Hand-maintained room grammar**: `apps/realtime/src/broadcast-audience.ts:17-32` enumerates every legitimate room shape "derived from src/index.ts socket.join calls" with hardcoded line references (`index.ts:656, 619-628, 698-699, 753, 821, 848, 940`). The actual joins now live at `index.ts:934-938, 966, 1010-1011, 1064, 1132, 1159` — **every cited line number is ~300 lines stale**. The shape list is a second encoding of the room grammar; adding a `socket.join` shape without updating the validator silently breaks broadcasts to it.

## Proposed direction

- Centralize revocation→kick at the permission mutation layer (one hook where permissions change, instead of per-route calls), so forgetting is impossible.
- Extract the room-name grammar into one module (builders + validator generated from the same definitions) shared by the join sites and `broadcast-audience`, and drop the line-number comments that can only rot.
- Decide the role of room membership honestly: if per-event reauth is the trust boundary (it is), room membership is a delivery optimization — document that, and consider whether the kick path can then be best-effort by design rather than semi-load-bearing.

---
Filed from the sources-of-truth audit.

### PR DESCRIPTION
**Pull request by 2witstudios**
## Summary

- Extracted the Socket.IO room-name grammar (builders + `isKnownRoomId` validator) into `@pagespace/lib/realtime/rooms` — the single source both the realtime join sites (`index.ts`) and the `broadcast-audience` validator now consume, instead of `broadcast-audience.ts`'s hand-maintained shape list whose `index.ts` line-number citations had drifted ~300 lines stale. A drift guard (`apps/realtime/src/__tests__/room-grammar-drift-guard.test.ts`) proves every builder output is a valid broadcast audience and that no realtime source hand-rolls a room name outside the shared module.
- Centralized revocation→kick as a hook at the permission-mutation layer (`packages/lib/src/permissions/revocation-kick.ts`): `revokePagePermission` now `await`s its own kick internally, and the remaining call sites that revoke access outside that function (drive member removal, activity rollback, page-goes-private) route through `kickForPagePermissionRevocation` / `kickForDriveMembershipRevocation` instead of hand-picking rooms per route — so a new revocation path can no longer forget to kick.
- Moved the kick transport (`kickUserFromRooms`) from `apps/web`'s `socket-utils.ts` into `@pagespace/lib/realtime/kick-client`, so the permission mutation layer can call it directly. `apps/realtime/src/kick-handler.ts` now imports `KickPayload`/`KickResult`/`KICK_REASONS` from that same module instead of hand-duplicating them.
- Documented room membership as a delivery optimization over the per-event permission recheck (`per-event-auth.ts`), which remains the authorization boundary; kicks are best-effort by design (never throw).
- `kickForDriveMembershipRevocation` fires the drive-scoped kicks (drive/calendar/activity — no page ID needed) concurrently with the drive's page-ID lookup, instead of gating them behind that DB round-trip.
- Dropped dead generality confirmed unused after centralization: the kick payload's prefix-wildcard support (`drive:*`) and two now-callerless wrapper functions in `kick-handler.ts`.

### Follow-ups from automated review passes
- **P1 (chatgpt-codex-connector):** `revokePagePermission` fired its revocation kick with `void` instead of `await`ing it, reopening the stale-delivery window the prior (awaited) per-route kicks had closed. Fixed to `await`; added a regression test proving the mutation's promise doesn't resolve while the kick is still in flight (verified RED against the `void` version first).
- **Proactive `/simplify` pass** (4 independent reuse/simplification/efficiency/altitude reviews over the full diff): fixed the concurrency issue above, removed the dead wildcard/wrapper code, deduplicated the kick payload types, simplified an `Extract<...>` to a plain literal type, and documented why the room grammar's `isCUID2` is intentionally *not* derived from the more permissive general-purpose ID validator (it's a security-boundary check matching the actual ID-generator invariant, not general ID parsing).

## Test plan
- [x] `bun run typecheck` (full monorepo, 16/16 tasks)
- [x] `packages/lib` full unit suite: 8988 passed
- [x] `apps/realtime` full suite: 967 passed, including the branch-coverage threshold (98%)
- [x] Targeted `apps/web` route/hook suites for all touched files: 229 passed
- [x] ESLint clean on all touched files
- [x] `bun run test:unit` (full monorepo) — 6 pre-existing failures unrelated to this change (missing test-DB Postgres role, ReDoS timing flakiness, a timezone-boundary test), none in touched files
- [x] CI green: Unit Tests, Lint & TypeScript, Security Test Suite, CodeQL

Closes #2158

https://claude.ai/code/session_01HPNkTkUtMo6Ce3nUdnMhMK

### CODE REVIEW

File: packages/lib/src/permissions/permission-mutations.ts [hunk 1]
```diff
@@ -416,6 +417,17 @@ export async function revokePagePermission(
     .delete(pagePermissions)
     .where(eq(pagePermissions.id, previousPermission.id));
 
+  // 5.5. Revocation→kick hook (#2158): evict the target's sockets from the
+  // page rooms here — at the mutation, not in each calling route — so no
+  // revocation path can forget it. Fire-and-forget like the audit log below:
+  // kicks are best-effort (room membership is a delivery optimization; the
+  // per-event recheck stays authoritative) and never throw.
+  void kickForPagePermissionRevocation({
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/permissions/permission-mutations.ts]: Good catch — fixed in 1e709ad64. `revokePagePermission` now `await`s `kickForPagePermissionRevocation` instead of firing it with `void`, restoring the same before-response guarantee the prior per-route `await Promise.all([kickUserFromPage(...), kickUserFromPageActivity(...)])` calls had. It still cannot fail the mutation — the hook is best-effort internally (Promise.allSettled over the underlying kicks) and never throws — so awaiting it only adds latency for the kick attempt to finish, not a new failure mode.

Also added a regression test (`permission-mutations-unit.test.ts`, "awaits the kick before resolving") that uses a manually-resolved deferred promise to prove `revokePagePermission`'s returned promise does not settle while the kick is still in flight. Verified it fails RED against the `void` version before confirming it passes with the `await` fix, so this specific regression cannot silently reappear.

Leaving this thread open for you/a reviewer to verify — not resolving it myself since I made the fix during this pass.
